gamemodes/terrortown/gamemode/server/sv_player.lua

```diff
@@ -511,88 +514,76 @@ function GM:KeyPress(ply, key)
--- @param Player ply The @{Player} pressing the key. If running client-side, this will always be @{LocalPlayer}
--- @param number key The key that the @{Player} pressed using <a href="https://wiki.facepunch.com/gmod/Enums/IN">IN_Enums</a>.
--- @hook
--- @realm server
--- @ref https://wiki.facepunch.com/gmod/GM:KeyRelease
--- @local
-function GM:KeyRelease(ply, key)
-    if key ~= IN_USE or not IsValid(ply) or not ply:IsTerror() then
-        return
-    end
-
-    -- see if we need to do some custom usekey overriding
-    local tr = util.TraceLine({
-        start = ply:GetShootPos(),
-        endpos = ply:GetShootPos() + ply:GetAimVector() * 100,
-        filter = ply,
-        mask = MASK_SHOT,
-    })
+-- Normally all dead players are blocked from "+use" on the server, meaning we
+-- can't let them search bodies. This sucks because searching bodies is
+-- fun. Hence on the client we override +use for specs and use this instead.
+local function SpecUseKey(ply, ent)
+    -- Do not allow the spectator to gather information if they're about to revive.
+    if ply:IsReviving() then
+        LANG.Msg(ply, "spec_about_to_revive", nil, MSG_MSTACK_WARN)
 
-    if not tr.Hit or not IsValid(tr.Entity) then
         return
     end
 
-    if tr.Entity.CanUseKey and tr.Entity.UseOverride then
-        local phys = tr.Entity:GetPhysicsObject()
-
-        if IsValid(phys) and not phys:HasGameFlag(FVPHYSICS_PLAYER_HELD) then
-            tr.Entity:UseOverride(ply)
-
-            return true
+    if ent.player_ragdoll then
+        if not ply:KeyDown(IN_WALK) then
+            CORPSE.ShowSearch(ply, ent)
         else
-            -- do nothing, can't +use held objects
-            return true
+            ply:Spectate(OBS_MODE_IN_EYE)
+            ply:SpectateEntity(ent)
         end
-    elseif tr.Entity.player_ragdoll then
-        CORPSE.ShowSearch(ply, tr.Entity, ply:KeyDown(IN_WALK) or ply:KeyDownLast(IN_WALK)) -- Body Corpse Search Identify TODO
-
-        return true
+    elseif ent:IsPlayer() and ent:IsActive() then
+        ply:Spectate(ply.spec_mode or OBS_MODE_IN_EYE)
+        ply:SpectateEntity(ent)
+    else
+        PROPSPEC.Target(ply, ent)
     end
 end
 
----
--- Normally all dead players are blocked from IN_USE on the server, meaning we
--- can't let them search bodies. This sucks because searching bodies is
--- fun. Hence on the client we override +use for specs and use this instead.
-local function SpecUseKey(ply, cmd, arg)
-    if not IsValid(ply) or not ply:IsSpec() then
+net.Receive("TTT2PlayerUseEntity", function(len, ply)
+    local ent = net.ReadEntity()
+
+    if not IsValid(ent) then
         return
     end
 
-    -- Do not allow the spectator to gather information if they're about to revive.
-    if ply:IsReviving() then
-        LANG.Msg(ply, "spec_about_to_revive", nil, MSG_MSTACK_WARN)
-
+    -- Check if the use interaction is possible
+    local distance = ply:GetShootPos():Distance(ent:GetPos())
+    if distance > 150 then
```

Comment:
in the game we use a use distance of 100 so far, so I think it should be 100 here as well. Also we should probably put the use distance in a global variable, maybe even a convar?

Reply:
The convar sounds like a good idea, but the server check here should be more than 100.

The client already checks that it's only in a distance of 100, but the server might be in a slightly different position and especially when the client aims at an outer position of an ent, but the server only compares to the center Position this can lead to the use being denied.
Not sure, if an additional trace should be done when range is over a hundred, but I thought this can make it more robust for the client and not lead to any behavior that can be abused a lot.

Reply:
But what happens if you look at a corpse? This already has a size of 160 I think, breaking this implementation because the center has a distance of 80 from the corners. Or even more if it is diagonal. I think you should use the EyeTrace on the server as well. Maybe add a 10% margin to it or so.

EyeTrace is also cached, so it shouldn't add any performance penalty.

However, now that I think about it: This would also break if the model is in a slightly different position on the server compared to the client. So maybe the entity shouldn't be checked at all on the server?

Reply:
The entity has to be checked on the server, otherwise you enable cheats like instant corpse reveal even if you are not close to it.

Instead of a trace you could maybe do a bounding box check, gotta look into that.
The eyetrace of course could be done, but it will only help in a few cases, where you are that far away, taht the check fails, but you are still close enough and looking at the server entity correctly. As this check was bad before, it would only help a little.

Reply:
Maybe this:
https://wiki.facepunch.com/gmod/util.IsBoxIntersectingSphere

Reply:
This here is perfect:
https://wiki.facepunch.com/gmod/Entity:BoundingRadius

You can just add it to the distance and it scales for all entities.
Flat/Boxed ragdoll => 45/40
C4 => 13
This will add the necessary distance to be safe.

Reply:
Perfect! This looks good to me then!